### PR TITLE
gh-122854: Add Py_HashBuffer() function

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-122854

----
📚 Documentation preview 📚: https://cpython-previews--122855.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
cc @pitrou @encukou

File: Doc/c-api/hash.rst [hunk 1]
```diff
@@ -89,6 +89,24 @@ See also the :c:member:`PyTypeObject.tp_hash` member and :ref:`numeric-hash`.
...
+
+.. c:function:: Py_hash_t Py_HashBytes(const void *ptr, Py_ssize_t len)
+
+   Compute and return the hash value of a buffer of *len* bytes
+   starting at address *ptr*. The hash is guaranteed to match that of
+   :class:`bytes`, :class:`memoryview`, and other built-in objects
+   that implement the :ref:`buffer protocol <bufferobjects>`.
+
+   Use this function to implement hashing for immutable objects whose
+   ``tp_richcompare`` function compares to another object's buffer.
```
Comment (erlend-aasland) [inline Doc/c-api/hash.rst]:
```suggestion
   :c:member:`~PyTypeObject.tp_richcompare` function compares to another object's buffer.
```

File: Doc/c-api/hash.rst [hunk 2]
```diff
@@ -89,6 +89,24 @@ See also the :c:member:`PyTypeObject.tp_hash` member and :ref:`numeric-hash`.
...
+
+   *len* must be greater than or to equal to ``0``.
+
+   The function cannot fail: it cannot return ``-1``.
```
Comment (erlend-aasland) [inline Doc/c-api/hash.rst]: We normally use this wording in the C API docs.

```suggestion
   This function always succeeds.
```

**Review by erlend-aasland — COMMENTED** [review]
Did you intend to update `Misc/stable_abi.toml` also?

File: Doc/c-api/hash.rst (hunk 2 above)
Comment (encukou) [reply Doc/c-api/hash.rst]: +1. We standardized on -1/NULL for *new* API, but we should not suggest that all C API uses that.

**Review by encukou — COMMENTED** [review]
Thank you! This looks good, but, wasn't the name [supposed to be `Py_HashBuffer`](https://github.com/capi-workgroup/decisions/issues/13#issuecomment-2025124073)?

File: Doc/c-api/hash.rst [hunk 3]
```diff
@@ -89,6 +89,24 @@ See also the :c:member:`PyTypeObject.tp_hash` member and :ref:`numeric-hash`.
...
+
+   Compute and return the hash value of a buffer of *len* bytes
+   starting at address *ptr*. The hash is guaranteed to match that of
+   :class:`bytes`, :class:`memoryview`, and other built-in objects
+   that implement the :ref:`buffer protocol <bufferobjects>`.
+
+   Use this function to implement hashing for immutable objects whose
+   ``tp_richcompare`` function compares to another object's buffer.
+
+   *len* must be greater than or to equal to ``0``.
```
Comment (picnixz) [inline Doc/c-api/hash.rst]:
```suggestion
   *len* must be greater than or equal to ``0``.
```

File: Objects/unicodeobject.c [hunk 4]
```diff
@@ -11672,7 +11672,7 @@ unicode_hash(PyObject *self)
     if (hash != -1) {
         return hash;
     }
-    x = _Py_HashBytes(PyUnicode_DATA(self),
+    x = Py_HashBytes(PyUnicode_DATA(self),
                       PyUnicode_GET_LENGTH(self) * PyUnicode_KIND(self));
```
Comment (picnixz) [inline Objects/unicodeobject.c]:
```suggestion
    x = Py_HashBytes(PyUnicode_DATA(self),
                     PyUnicode_GET_LENGTH(self) * PyUnicode_KIND(self));
```

File: Lib/test/test_capi/test_hash.py [hunk 5]
```diff
@@ -77,3 +77,13 @@ def python_hash_pointer(x):
         # Py_HashPointer((void*)(uintptr_t)-1) doesn't return -1 but -2
         VOID_P_MAX = -1 & (2 ** (8 * SIZEOF_VOID_P) - 1)
         self.assertEqual(hash_pointer(VOID_P_MAX), -2)
+
+    def test_hash_bytes(self):
+        hash_bytes = _testcapi.hash_bytes
+
+        def check(data):
```
Comment (picnixz) [inline Lib/test/test_capi/test_hash.py:84]: Should we add a test where the input is of incorrect type?

Comment (vstinner) [reply Lib/test/test_capi/test_hash.py:84]: `Py_HashBytes(const void *ptr, Py_ssize_t len)` is a C API which accepts any memory, it has no Python type.

Comment (picnixz) [reply Lib/test/test_capi/test_hash.py:84]: Oh right. Sorry, I think I got confused by

```rst
The hash is guaranteed to match that of
:class:`bytes`, :class:`memoryview`, and other built-in objects
that implement the :ref:`buffer protocol <bufferobjects>`.
```

(The fault lies with me, the docs have no issues).

File: Modules/_testcapi/hash.c [hunk 6]
```diff
@@ -64,14 +85,14 @@ object_generichash(PyObject *Py_UNUSED(module), PyObject *arg)
...
-    Py_BUILD_ASSERT(sizeof(long long) >= sizeof(hash));
-    return PyLong_FromLongLong(hash);
+    return long_from_hash(hash);
 }
 
 
 static PyMethodDef test_methods[] = {
     {"hash_getfuncdef", hash_getfuncdef, METH_NOARGS},
     {"hash_pointer", hash_pointer, METH_O},
+    {"hash_bytes", hash_bytes, METH_VARARGS},
```
Comment (encukou) [inline Modules/_testcapi/hash.c]: Nitpick: This should probably be `hash_buffer` too.

**Review by encukou — COMMENTED** [review]
With Erlend's suggestions, this should be good to merge!

**Comment by vstinner** [comment]
> With Erlend's suggestions, this should be good to merge!

Fixed. Oh, I thought that I already addressed his review, but it was maybe on another similar PR.

I also replaced hash_bytes() with hash_buffer() in tests. Nicely spotted @encukou.

**Comment by vstinner** [comment]
@erlend-aasland:
> Did you intend to update Misc/stable_abi.toml also?

https://github.com/capi-workgroup/decisions/issues/13 doesn't mention the stable ABI. Do you think that this API should be added to the stable ABI?

**Comment by vstinner** [comment]
@erlend-aasland: I enabled auto-merge. If you consider that the function should be added to the limited C API 3.14, I can prepare a follow-up PR.

**Comment by vstinner** [comment]
Merged, thanks for reviews.

**Pull request merged**